File: mylib/union_find.py
class UnionFind:
    def __init__(self, V):
        self.parent = {}
        self.rank = collections.defaultdict(int)
        for v in V:
            self.parent[v] = v
            self.rank[v] = 0

    def find(self, x):
        #if x == -1:
        #    return -1

        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])   
        return self.parent[x] 
    
    def isSameSet(self, x, y):
        return self.find(x) == self.find(y)

    def union(self, x, y):
        r1 = self.find(x)
        r2 = self.find(y)
        if r1 == r2:
            return False   # union fail: u v already belong to the same set   
        if self.rank[r1] > self.rank[r2]:
            self.parent[r2] = r1
        elif self.rank[r1] < self.rank[r2]:
            self.parent[r1] = r2
        else:
            self.rank[r1]+=1
            self.parent[r2] = r1
        
        return True
    
    def sizeOfSet(self):
        for v in self.V:
            self.find(v)
        C = collections.Counter(self.parent.values())
        return len(C.keys())

File: mylib/union_find.cpp
#include <vector>
#include <set>
using namespace std;


struct UnionFind {
    UnionFind(int n):pset(n) {
        for (int i = 0; i < n; i++)
            pset[i]= i;
    };
    int find (int x) {
        if (pset[x] !=x)
            pset[x] = find(pset[x]);
        
        return pset[x];
    }
    
    bool Union(int x, int y) {
        int px = find(x);
        int py = find(y);
        if (px == py)
            return false;
        pset[px] = py;
        return true;
    }
    
    int numOfComponents() {
        for(int i= 0; i < pset.size(); i++)
            find(i);
        
        set<int> s(pset.begin(), pset.end());
        return s.size();
    }
    
    vector<int> pset;
};